README.md
# grunt-vnuserver-dev

[![Dependency Status](https://img.shields.io/david/prantlf/grunt-vnuserver.svg)](https://david-dm.org/prantlf/grunt-vnuserver)
[![devDependency Status](https://img.shields.io/david/dev/prantlf/grunt-vnuserver.svg)](https://david-dm.org/prantlf/grunt-vnuserver#info=devDependencies)

[Grunt][grunt] plugin for starting the [vnu.jar markup checker][vnujar] in server mode.
Plays well with [grunt-html-dev][grunt-html-dev] for faster HTML validation by only starting vnu.jar once, as startup can take a few seconds.

This fork publishes a new `grunt-vnuserver-dev` NPM module, which depends on `vnu-jar@next` instead of the latest release of `vnu-jar`. The [W3C Markup Validation Service](https://validator.w3.org/) uses the *development version* of `vnu-jar` too. If you want to get consistent results from on-line and off-line testing, you should use `grunt-vnuserver-dev` instead of `grunt-vnuserver` in your project.

## Getting Started

Install this grunt plugin next to your project's [Gruntfile.js][getting_started] with:

```bash
npm install grunt-vnuserver-dev --save-dev
```

Then add this line to your project's `Gruntfile.js`:

```js
grunt.loadNpmTasks('grunt-vnuserver-dev');
```

## Options

### `port`

* Type: `Number`
* Default: 8888

The port on which to start the server.

```js
all: {
  options: {
      port: 8877
  },
}
```

### `skippable`

* Type: `Boolean`
* Default: `false`

Whether or not to skip server startup if port is already in use.
Task will fail if port is in use and `skippable` and `useAvailablePort` are false.

### `persist`

* Type: `Boolean`
* Default: `false`

Whether or not to keep the vnu server running even after grunt exists.
If false, vnu server is killed when grunt exists.

### `useAvailablePort`

* Type: `Boolean`
* Default: `false`

If `true` the task will look for the next available port, if the port
set by the `port` option is in use.

## Example

Consider the following configuration in Gruntfile.js in which the `watch` task is set to run `htmllint` every time the source file changes.
By starting the validator in server mode once using the `vnuserver` task, validations by `htmllint` can be performed much faster by simply connecting to this already-running server.

```js
module.exports = function (grunt) {
    grunt.initConfig({
        vnuserver: {
        },
        htmllint: {
            all: {
                options: {
                    // This option makes grunt-html-dev connect to the vnu server instance.
                    server: {}
                },
                src: "app.html"
            }
        },
        watch: {
            all: {
                tasks: ['htmllint'],
                files: "app.html"
            }
        },
    });

    grunt.loadNpmTasks('grunt-vnuserver-dev');
    grunt.loadNpmTasks('grunt-html-dev');
    grunt.loadNpmTasks('grunt-contrib-watch');

    grunt.registerTask('default', ['vnuserver', 'watch']);
};
```

### Using the First Available Port

If you set `useAvailablePort` to `true`, you will need to pass the actual value to the `htmllint`
task, but the value will be known first during the runtime. Use a function for the `server`
option, which was introduced in the [grunt-html-dev] fork, at first.

```js
module.exports = function (grunt) {
    var vnuPort;

    grunt.initConfig({
        vnuserver: {
            // Name the task to be able to listen to its events.
            options: {
                // Start with the first free ephemeral port.
                port: 49152,
                // Try other ports, up to port + 30, if the first one is not free.
                useAvailablePort: true
            }
        },
        htmllint: {
            options: {
                // Connect to the vnu server on the dynamically chosen  port.
                server: function () {
                    return {
                        port: vnuPort
                    };
                }
            },
            all: {
                src: "app.html"
            }
        },
        watch: {
            all: {
                tasks: ['htmllint'],
                files: "app.html"
            }
        },
    });

    grunt.loadNpmTasks('grunt-vnuserver-dev');
    grunt.loadNpmTasks('grunt-html-dev');
    grunt.loadNpmTasks('grunt-contrib-watch');

    // Obtain the port, which the vnu server is listening to.
    grunt.event.on('vnuserver.listening', function (port) {
        vnuPort = port;
        // vnuPort = grunt.config.get('vnuserver.options.port');
    });

    grunt.registerTask('default', ['vnuserver', 'watch']);
};
```

## License

Copyright Bennie Swart.
Licensed under the MIT license.

[grunt]: http://gruntjs.com/
[grunt-html-dev]: https://github.com/prantlf/grunt-html-dev
[getting_started]: http://gruntjs.com/getting-started
[vnujar]: https://validator.github.io/validator/


CHANGELOG.md
# [2.0.0](https://github.com/prantlf/grunt-vnuserver/compare/v1.2.0...v2.0.0) (2020-01-10)

### Bug Fixes

* Detect that the new vnu-jar server has started using log4j ([dffd7e8](https://github.com/prantlf/grunt-vnuserver/commit/dffd7e80fe4e61200baf8b8eb0692193ff6aa623))
* Upgrade package dependencies ([288c8fc](https://github.com/prantlf/grunt-vnuserver/commit/288c8fc402f8889c3cf51bca8744b587758fae2c))

### BREAKING CHANGES

* The minimum required Node.js version is the latest release of the 8 branch. Most dependencies require this now.

# [1.2.0](https://github.com/prantlf/grunt-vnuserver/compare/v1.1.3...v1.2.0) (2018-09-08)

### Features

* Remove the check for supported Java versions

The current one is Java 10 and OpenJDK changed versioning to 10.x. The version check has become troublesome to constantly update. It is better to document what Java version work or do not work with this package.

## [1.1.3](https://github.com/prantlf/grunt-vnuserver/compare/v1.1.2...v1.1.3) (2018-06-18)

### Bug Fixes

* Switch to vnu-jar@next, which is the renamed [@dev](https://github.com/dev) package ([f86f8e6](https://github.com/prantlf/grunt-vnuserver/commit/f86f8e6a143a387fa017a0899cff2c523db272c1))

## [1.1.2](https://github.com/prantlf/grunt-vnuserver/compare/v1.1.1...v1.1.2) (2018-06-18)

### Bug Fixes

* Switch to the versioned vnu-jar because the [@dev](https://github.com/dev) package was removed ([faf7502](https://github.com/prantlf/grunt-vnuserver/commit/faf75026b23f1c119d849cbcae1af46c11c290ca))

This is the first release of the [`grunt-vnuserver-dev` module](https://www.npmjs.com/package/grunt-vnuserver-dev) forked from the [original project](https://github.com/bennieswart/grunt-vnuserver).


Gruntfile.js
'use strict'

module.exports = function (grunt) {
    const coverage = process.env.GRUNT_VNUSERVER_COVERAGE;

    grunt.initConfig({
        eslint: {
            check: [
                'Gruntfile.js',
                'tasks/*.js'
            ]
        },

        instrument: {
            files: 'tasks/*.js',
            options: {
                lazy: true,
                basePath: 'coverage/'
            }
        },

        storeCoverage: {
            options: {
                dir: 'coverage'
            }
        },

        makeReport: {
            src: 'coverage/coverage.json',
            options: {
                type: 'lcov',
                dir: 'coverage',
                print: 'detail'
            }
        },

        clean: {
            coverage: ['coverage/**']
        },

        vnuserver: {
            options: {
                port: 49152,
                useAvailablePort: true
            }
        }
    });

    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-eslint');
    grunt.loadNpmTasks('grunt-istanbul');
    grunt.loadTasks(coverage ? 'coverage/tasks' : 'tasks');

    let vnuPort;

    grunt.event.on('vnuserver.listening', function (port) {
        vnuPort = port;
    });

    grunt.registerTask('vnucheck', function () {
        if (!vnuPort) {
            grunt.fail.warn('VNU Server has not started.');
        }
    });

    const test = ['eslint', 'clean', 'vnuserver', 'vnucheck'];
    const report = coverage ? ['storeCoverage', 'makeReport'] : [];
    grunt.registerTask('default', test.concat(report));
};


tasks/vnuserver.js
let javadetect = require('grunt-html-dev/lib/javadetect');
let jar = require('vnu-jar');
let path = require('path');
let portastic = require('portastic');

let MAX_PORTS = 30;

module.exports = function (grunt) {
    grunt.registerTask('vnuserver', 'Start the Nu Html Checker server.', function () {
        let opt = this.options({port: 8888, skippable: false, persist: false, useAvailablePort: false});
        let done = this.async();

        start(opt.port);

        function start (port) {
            portastic.test(port, function (open) {
                if (!open) {
                    if (opt.useAvailablePort) {
                        if (port < opt.port + MAX_PORTS) {
                            grunt.log.debug('Port ' + port + ' in use. Trying another one.');
                            process.nextTick(function () {
                                start(port + 1);
                            });
                        } else {
                            done(Error('Port ' + opt.port + ' and ' + MAX_PORTS + ' others in use. To ignore, set skippable: false.'));
                        }
                    } else if (opt.skippable) {
                        grunt.log.debug('Port ' + port + ' in use. Skipping server startup.');
                        done();
                    } else {
                        done(Error('Port ' + port + ' in use. To ignore, set skippable: false.'));
                    }
                    return;
                }

                grunt.config.set('vnuserver.options.port', port);
                grunt.event.emit('vnuserver.listening', port);

                let child;
                let cleanup = function () {
                    let killing = grunt.log.write('Killing vnuserver...');
                    child.kill('SIGKILL');
                    killing.ok();
                };
                if (!opt.persist) {
                    process.on('exit', cleanup);
                    let exit = grunt.util.exit;
                    grunt.util.exit = function () { // This seems to be the only reliable on-exit hook.
                        cleanup();
                        return exit.apply(grunt.util, arguments);
                    };
                }

                javadetect(function(err, java) {
                    if (err) {
                        throw err;
                    }
                    let args = [
                      (java.arch === 'ia32' ? '-Xss512k' : ''),
                      '-Dnu.validator.servlet.log4j-properties=' +
                        path.join(__dirname, 'log4j.properties'),
                      '-Dnu.validator.servlet.read-local-log4j-properties=1',
                      '-cp', jar, 'nu.validator.servlet.Main', port
                    ].filter(x => x);
                    let vnustartup = grunt.log.write('Starting vnuserver on port ' + port + '...');
                    child = grunt.util.spawn({cmd: 'java', args: args}, function(error /*, stdout, stderr*/) {
                        if (error && (error.code !== 1 || error.killed || error.signal)) {
                            done(false);
                        }
                    });
                    child.stderr.on('data', function (chunk) {
                        if (chunk.toString().indexOf('org.eclipse.jetty.server.Server: Started') >= 0) {
                            vnustartup.ok();
                            done();
                        }
                        if (chunk.toString().indexOf('java.net.BindException: Address already in use') >= 0) {
                            vnustartup.error();
                            done(Error('Port ' + port + ' in use. Shutting down.'));
                            cleanup();
                        }
                    });
                });
            });
        }
    });
};
